presto-main/src/main/java/com/facebook/presto/tracing/SimpleTracer.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.tracing.Tracer;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static com.facebook.presto.spi.StandardErrorCode.DISTRIBUTED_TRACING_ERROR;
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+public class SimpleTracer
+        implements Tracer
+{
+    public final Map<String, SimpleTracerBlock> blockMap = new ConcurrentHashMap<>();
+    public final Map<String, SimpleTracerBlock> recorderBlockMap = new LinkedHashMap<>();
+    public final List<SimpleTracerPoint> pointList = new CopyOnWriteArrayList<>();
+
+    public SimpleTracer()
+    {
+        addPoint("Start tracing");
+    }
+
+    @Override
+    public void addPoint(String annotation)
+    {
+        pointList.add(new SimpleTracerPoint(annotation));
+    }
+
+    @Override
+    public void startBlock(String blockName, String annotation)
+    {
+        if (blockMap.containsKey(blockName)) {
+            throw new PrestoException(DISTRIBUTED_TRACING_ERROR, "Duplicated block inserted: " + blockName);
+        }
+        SimpleTracerBlock block = new SimpleTracerBlock(annotation);
+        blockMap.put(blockName, block);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **49 and 53**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **55**. If the key already exists, the code should throw an error at line: **50**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.